feat(function): new stuff (#40)
amber-lang/amber-docs: docs/basic_syntax/functions.md

An interesting fact about functions is that they are not parsed unless they are used. This behavior exists because Amber allows you to omit specifying any type at all. When you use such function - then it generates different variants of this function with types that were used (without any duplications).

You can use default parameters but requires to define the type of the parameter:

```ab
fun addition(a: Num, b: Num = 100): Num {
    return a + b

> }

Reviewer: ```suggestion
}

```

> echo addition(10)
> // Outputs: 110

Reviewer: ```suggestion
echo addition(10) // Outputs: 110
echo addition(10, 20) // Outputs: 30
```

Notice that using `?` operator is automatically failing with the `status` code of the failing operation.

> You can mark a whole function as failable using `?` in the function declaration, an example:

Reviewer: ```suggestion
If you specify the return type of a failable function, you must also append the `?` to the type name.
```

```ab
fun failable(): Num? {
    if 0 > 5 {
        fail 1
    }

    return 1
}

> let a = failable() failed: echo "Failed"
> 
> if a is Num: echo "Succeded"

Reviewer: ```suggestion
```

Author: why do you want to remove those lines?

Reviewer: They are unnecessary to the explanation and diverge attention.

Author: In my opinion is a way to show how to use the function also if failable

Reviewer: That's totally unrelated to the discussed topic, though. The way failable functions are used is already demonstrated separately.

```

> 

Reviewer: ```suggestion

Note that you cannot force a function to become failable by simply appending the `?` to the return type. The `?` can (and must) only be used in a function declaration, if the function is known to be failable.
```

Notice that using `?` operator is automatically failing with the `status` code of the failing operation.

If you specify the return type of a failable function, you must also append the `?` to the type name.

```ab
fun failable(): Num? {
    if 0 > 5 {
        fail 1
    }

    return 1
}

> 

Reviewer: ```suggestion
```
A leftover newline, should be removed. Otherwise LGTM.
